toolchain/semantics/semantics_ir_formatter.cpp

```diff
@@ -0,0 +1,654 @@
+        llvm::StringMapEntry<NameResult>* value = value_;
+        CARBON_CHECK(value) << "cannot print a null name";
+        while (value->second.ambiguous && value->second.fallback) {
+          value = value->second.fallback.value_;
+        }
+        return value->first();
+      }
+
+      operator llvm::StringRef() const { return str(); }
+      operator llvm::Twine() const { return str(); }
+      operator std::string() const { return str().str(); }
+
+      auto SetFallback(Name name) -> void { value_->second.fallback = name; }
+
+      auto SetAmbiguous() -> void { value_->second.ambiguous = true; }
+
+     private:
+      llvm::StringMapEntry<NameResult>* value_;
+    };
+
+    struct NameResult {
+      bool ambiguous = false;
+      Name fallback = Name();
+    };
+
+    llvm::StringRef prefix;
+    llvm::StringMap<NameResult> allocated = {};
+    int unnamed_count = 0;
+
+    auto AddNameUnchecked(llvm::StringRef name) -> Name {
+      return Name(allocated.insert({name, NameResult()}).first);
+    }
+
+    auto AllocateName(const NodeNamer& namer, ParseTree::Node node,
+                      std::string name = "") -> Name {
+      // The best (shortest) name for this node so far, and the current name
+      // for it.
+      Name best;
+      Name current;
+
+      // Add `name` as a name for this entity.
+      auto add_name = [&](bool mark_ambiguous = true) {
+        auto [it, added] = allocated.insert({name, NameResult()});
+        Name new_name = Name(it);
+
+        if (!added) {
+          if (mark_ambiguous) {
+            // This name was allocated for a different node. Mark it as
+            // ambiguous and keep looking for a name for this node.
+            new_name.SetAmbiguous();
+          }
+        } else {
+          if (!best) {
+            best = new_name;
+          } else {
+            current.SetFallback(new_name);
+          }
+          current = new_name;
+        }
+        return added;
+      };
+
+      // All names start with the prefix.
+      name.insert(0, prefix.data(), prefix.size());
+
+      // Use the given name if it's available and not just the prefix.
+      if (name.size() > prefix.size()) {
+        add_name();
+      }
+
+      // Append location information to try to disambiguate.
+      if (node.is_valid()) {
+        auto token = namer.parse_tree_.node_token(node);
+        name += ".loc";
+        name += llvm::itostr(namer.tokenized_buffer_.GetLineNumber(token));
+        add_name();
+
+        name += "_";
+        name += llvm::itostr(namer.tokenized_buffer_.GetColumnNumber(token));
+        add_name();
+      }
+
+      // Append numbers until we find an available name.
+      name += ".";
+      auto name_size_without_counter = name.size();
+      for (int counter = 1;; ++counter) {
+        name.resize(name_size_without_counter);
+        name += llvm::itostr(counter);
+        if (add_name(/*mark_ambiguous=*/false)) {
+          return best;
+        }
+      }
```

Comment:
Would using a string-stream be clearer? It seems a tiny bit better to me, but optional if you prefer as-is:

```suggestion
      // Append location information to try to disambiguate.
      llvm::raw_string_ostream name_os(name);
      if (node.is_valid()) {
        auto token = namer.parse_tree_.node_token(node);
        name_os << ".loc" << namer.tokenized_buffer_.GetLineNumber(token);
        add_name();

        name_os << "_" << namer.tokenized_buffer_.GetColumnNumber(token);
        add_name();
      }

      // Append numbers until we find an available name.
      name_os << ".";
      auto name_size_without_counter = name.size();
      for (int counter = 1;; ++counter) {
        name.resize(name_size_without_counter);
        name_os << counter;
        if (add_name(/*mark_ambiguous=*/false)) {
          return best;
        }
      }
```

Reply:
I'm uncomfortable about modifying the `std::string` while it's being held by the ostream (even though it looks like `raw_string_ostream` doesn't cache anything about the string between output calls right now).

Done, creating the stream more frequently.